backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        {
+            _options = options.Value;
+        }
+        #endregion
+
+        #region Methods
+        /// <summary>
+        /// Get all of the project reports.
+        /// </summary>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view reports.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectReport> GetAll()
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var snapshots = this.Context.ProjectReports;
+
+            return snapshots;
+        }
+
+        /// <summary>
+        /// Get the project report for the specified 'id'.
+        /// </summary>
+        /// <param name="reportId"></param>
+        /// <exception cref="ArgumentNullException">Argument 'reportId' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view report.</exception>
+        /// <returns></returns>
+        public ProjectReport Get(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var report = this.Context.ProjectReports
+                .FirstOrDefault(p => p.Id == reportId) ?? throw new KeyNotFoundException();
+
+            return report;
+        }
+
+        /// <summary>
+        /// Get all project snapshots matching either the 'To' or 'From' dates of the passed report.
+        /// </summary>
+        /// <param name="reportId"></param>
+        /// <exception cref="ArgumentNullException">Argument 'reportId' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var snapshots = this.Context.ProjectSnapshots
+                .Include(s => s.Project).ThenInclude(p => p.Agency)
+                .Include(s => s.Project).ThenInclude(p => p.Status)
+                .Where(s => s.SnapshotOn == report.From || s.SnapshotOn == report.To)
+                .ToArray();
+
+            foreach(ProjectSnapshot snapshot in snapshots)
+            {
+                var previousSnapshot = snapshots.Where(s => s.ProjectId == snapshot.ProjectId && s.Id != snapshot.Id).FirstOrDefault();
+                snapshot.BaselineIntegrity = (snapshot?.NetProceeds ?? 0) - (previousSnapshot?.NetProceeds ?? 0);
+            }
+
+            return snapshots;
+        }
+
+        /// <summary>
+        /// Get all project snapshots matching either the 'To' or 'From' dates of the passed report.
```

Comment:
Is it? this calls function the above function.

Reply:
I think it should be
Get all project snapshots for the specified `reportId`.  Depending on whether you introduce saving the `Variance` to the database, it would only actually need to return the snapshots that match the `To` date.

Reply:
Ok, this may be something else that I misinterpreted from your design. Based on this:
https://preview.uxpin.com/bcfcebd6d46b70cc4030fc7ec0b363b72d9d2d99#/pages/133385268/simulate/sitemap
I assumed that I always need to return the To: and From: snapshots for every matching project. This results in two snapshots per project, each on a separate row, as in the linked design.

I can easily change this to only return the To: snapshots. I just thought some of the other differences might be useful as well. IE, if the netbook changed between snapshots.

Reply:
The UI only needs the `To` snapshots + variances.
However depending on whether you recalculate each time.  If you recalculate you would need to load the `From` snapshots before returning the results and doing the variance math.